episodes/escape-from-jupyter.md

```diff
@@ -0,0 +1,309 @@
+- **Track changes:** Using .py files and modules makes it easy for to see line-by-line
+changes you or others make to files.
+- **Concretize final code:** While you might test out three versions of a `transform_eia_gen_fuel()`
+function in a Jupyter notebook or make four exploratory plots, you'll ultimately want to make sure
+you're running the code you need for your final transformation process, and *only* that code. Moving
+to modules helps us distinguish between our exploration process and our final code.
+- **Reuse code:** Rather than copy-pasting a useful snippet or function into each notebook you're working in,
+you can store essential functions in one place and reuse them across your code, just like you'd import any other
+Python package.
+- **Test your code:** We'll cover this next! In short, using scripts and modules unlocks a world of tools
+you can use to test, debug and correct your code.
+
+### Creating a codebase
+
+What changes when we move our code out of a Jupyter notebook? One of the first roadblocks
+to creating a codebase is specifying which packages need to be installed to run your code.
+We have to make sure that collaborators have the same packages *and* versions of those
+packages installed to avoid unexpected problems.
+
+Luckily for us, developers have made it possible to set up a **virtual environment**
+in which to run any code we write. A virtual environment is a box that you can use to
+wrap up your project and hand it over to a collaborator - it tells their computer how to
+replicate the environment you used when developing your code (e.g., which packages, which Python version).
+
+Like other tools you may have encountered (`pip`, `pyenv`, `virtualenv`), [`uv`](https://docs.astral.sh/uv/) is a tool that helps you install and update packages, and then share those exact installation
+instructions with your peers. In fact, if you've run any of the code in prior episodes,
+you're already using `uv`! Helpfully, as we move away from Jupyter, we can use `uv` to set up a
+skeleton for our code project.
+
+:::: callout
+If you haven't yet installed `uv`, follow the [setup instructions](../learners/setup.md) before continuing.
+::::
+
+Open up your shell (see here for [OS-specific instructions](https://swcarpentry.github.io/shell-novice/index.html#open-a-new-shell).) In a terminal, navigate up one folder.
+```shell
+cd ..
+```
+
+Let's pick a short but descriptive name for our project (avoid using spaces): `pr-gen-fuel`.
+
+Now, run:
+```shell
+uv init pr-gen-fuel
+```
+
+What happened? If we navigate to the new folder that has been created, we can see a series
+of new files.
+
+```shell
+cd pr-gen-fuel
+ls
+```
+
+```
+main.py  pyproject.toml  README.md
+```
+
+We'll talk through each file in a moment, but first, let's just try and run our `.py` file.
+Rather than using `python main.py`, we use `uv run` to run the script within our virtual
+environment.
```

Comment:
```suggestion
We will use `uv run` to run the script within our virtual
environment.
```

Reply:
(if we need a way to ward off people who have used `python main.py` before, we can do that in a callout)